packages/opentelemetry-core/src/trace/Span.ts

```diff
@@ -0,0 +1,167 @@
+  private _attributes: types.Attributes = {};
+  private _links: types.Link[] = [];
+  private _events: types.Event[] = [];
+  private _status: types.Status = {
+    code: types.CanonicalCode.OK,
+  };
+  private _ended = false;
+  private _startTime: number;
+  private _duration = 0;
+
+  /** Constructs a new Span instance. */
+  constructor(
+    readonly _parentTracer: types.Tracer,
+    readonly _spanName: string,
+    readonly _options: types.SpanOptions
+  ) {
+    this._name = _spanName;
+    this._kind = _options.kind || types.SpanKind.INTERNAL;
+    this._spanId = randomSpanId();
+    const spanContext = this._getSpanContext(_options.parent);
+    if (spanContext && isValid(spanContext)) {
+      // New child span.
+      this._traceId = spanContext.traceId;
+      this._parentId = spanContext.spanId;
+      this._traceState = spanContext.traceState;
+    } else {
+      // This is a root span so no remote or local parent.
+      this._traceId = randomTraceId();
+    }
+    this._startTime = _options.startTime || performance.now();
+  }
+
+  tracer(): types.Tracer {
+    return this._parentTracer;
+  }
+
+  context(): types.SpanContext {
+    const spanContext: types.SpanContext = {
+      traceId: this._traceId,
+      spanId: this._spanId,
+      traceOptions: types.TraceOptions.SAMPLED,
+    };
+    if (this._traceState) {
+      spanContext.traceState = this._traceState;
+    }
+    return spanContext;
+  }
+
+  setAttribute(key: string, value: unknown): this {
+    if (this._isSpanEnded()) return this;
+
+    try {
+      // If the value is a typeof object it has to be JSON.stringify-able
+      const serializedValue =
+        typeof value === 'object' ? JSON.stringify(value) : value;
+      this._attributes[key] = serializedValue;
+    } catch (e) {
+      // @todo: log error
+    }
+    return this;
+  }
+
+  setAttributes(attributes: types.Attributes): this {
+    Object.keys(attributes).forEach(key => {
+      this.setAttribute(key, attributes[key]);
+    });
+    return this;
+  }
+
+  addEvent(name: string, attributes?: types.Attributes): this {
+    if (this._isSpanEnded()) return this;
+    this._events.push({ name, attributes });
+    return this;
+  }
+
+  addLink(spanContext: types.SpanContext, attributes?: types.Attributes): this {
+    if (this._isSpanEnded()) return this;
+    this._links.push({ spanContext, attributes });
+    return this;
+  }
+
+  setStatus(status: types.Status): this {
+    if (this._isSpanEnded()) return this;
+    this._status = status;
+    return this;
+  }
+
+  updateName(name: string): this {
+    if (this._isSpanEnded()) return this;
+    this._name = name;
+    return this;
+  }
+
+  end(endTime?: number): void {
+    if (this._isSpanEnded()) return;
+    this._ended = true;
+    const finishTime = endTime || performance.now();
+    this._duration = finishTime - this._startTime;
+    // @todo: record or export the span
+  }
+
+  isRecordingEvents(): boolean {
```

Comment:
can the span be recording events if the span/trace isn't sampled? What if the span is sampled but isRecordingEvents was not set in the options?

Reply:
I think the assumption is all the sampled span has `isRecordingEvents` flag as true.

Java's implementation: https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java#L179